### DISCUSSION TITLE
need help to fix a problem with a corrupted version of node nvm

### DISCUSSION THREAD

**Discussion by aleques-r** [coreybutler/nvm-windows#959] [Support]
I researched a lot on the subject, but I couldn't find a viable solution to my problem, I'm setting up a development environment by NVM I installed node 8.9.0, I managed to install the NPM package on it and the grunt libraries I needed, however, now I'm trying to do the same procedure, but in version 4.4.0 and it doesn't work, any command I give in version 4.4.0 doesn't return any errors

I already tried all the steps in this tutorial: [How to completely remove node.js from Windows](https://stackoverflow.com/questions/20711240/how-to-completely-remove-node-js-from-windows) but it didn't solve my problem either

on another clean machine I installed node 4.4.0 before version 8.9.0 and it worked normally, but it seems that having installed version 8 before version 4 generated this problem

problem image link: https://ibb.co/SQjR9Wz

**Comment by coreybutler** [comment]
The instructions you're following are for a standard installation of Node.js, not an NVM4W installation of Node.js. When you run `nvm use x.x.x`, you get a clean npm environment. 

Those versions of node are old enough that the current grunt modules may not be supported in those old versions. You may need to install an older version of grunt (or any other module) that supports Node 4.4.0 or 8.9.0 respectively. Remember, both of those versions reached end-of-life a long time ago. Most module authors stopped supporting those versions and often force updates to mitigate security vulnerabilities... but it's entirely up to the module maintainer.

The main point here is it is not clear whether there is a problem with NVM4W configuring Node or (more likely) if there is a problem with the module or the very old version of npm. When you run `npm <anything>`, it's entirely npm and node running... NVM4W doesn't touch that command, so it's just running an old version of npm that may not output the error messages people are used to seeing with newer versions.

**Reply by aleques-r** [reply]
well, I initially installed NVM 1.1.10 and installed node version 8.9.0, then ran:
npm install grunt -g
npm install grunt-cli -g
npm install graceful-fs -g
npm install grunt-ng-annotate -g

no problems with commands and packages, then I tried to install node 4.4.0 and the same packages and then the error in question appears...

On another computer without ever having used NVM I installed it, then I first installed node 4.4.0 and the packages and I don't have any errors, everything works normally
I don't know what to do, I'm almost thinking about formatting the machine and configuring it from scratch starting with 4.4.0 rs

**Reply by coreybutler** [reply]
I've attempted this in my local environment.

**Install Node 8.9.0**
<img width="1109" alt="image" src="https://user-images.githubusercontent.com/770982/231618872-cd36d054-8609-4129-ac34-967b2363b22e.png">

**Install Grunt Globally**
<img width="1110" alt="image" src="https://user-images.githubusercontent.com/770982/231618958-47c075fd-2430-4eb1-94e4-22e7cb3013c0.png">

**Install Grunt CLI Globally**
<img width="1110" alt="image" src="https://user-images.githubusercontent.com/770982/231619051-a66b0846-d4c3-4210-b3ad-8fc60b2bfa9b.png">

**Install graceful-fs Globally**
<img width="1109" alt="image" src="https://user-images.githubusercontent.com/770982/231619113-8a003a88-fd4d-4d15-a301-ea0767cf14c3.png">

**Install grunt-ng-annotate Globally**
<img width="1111" alt="image" src="https://user-images.githubusercontent.com/770982/231619172-6502cd8c-5aa8-4500-836d-63b555606712.png">

For good measure, I ran the grunt directly in the CLI and got the expected results:
<img width="1107" alt="image" src="https://user-images.githubusercontent.com/770982/231619253-8bd00ed3-8a7f-469e-8de0-e6f2e73e1556.png">

**Install and Use Node 4.4.0**
<img width="1109" alt="image" src="https://user-images.githubusercontent.com/770982/231619337-1e2c9d52-374e-41d5-93bd-eee3c09fb8a4.png">

**Install Grunt Globally**
<img width="1107" alt="image" src="https://user-images.githubusercontent.com/770982/231619552-6f5535e6-e110-4555-958a-40a9ddfa06d6.png">

As I mentioned initially, the grunt package expects a newer version of Node.js. It will still install, but that doesn't mean it will run properly.

Similar story for grunt-cli and grunt-ng-annotate:
<img width="1108" alt="image" src="https://user-images.githubusercontent.com/770982/231619836-580318bb-6076-4cfc-ad84-9adbed71d5fc.png">

<img width="1109" alt="image" src="https://user-images.githubusercontent.com/770982/231619956-5d63d48b-9ad9-4a6d-abf1-c1c0b2448f6d.png">

`graceful-fs` was fine:
<img width="1108" alt="image" src="https://user-images.githubusercontent.com/770982/231620052-c440b1b0-465f-428a-9747-bfcc141aa4b9.png">

Also, as expected, running grunt directly fails because it uses spread operator, which was not supported in Node 4.4.0.
<img width="1108" alt="image" src="https://user-images.githubusercontent.com/770982/231620195-79afd7e0-4242-4f7a-bf1e-f51cbdac473c.png">

NVM for Windows is not the culprit here. All NVM4W does is install Node.js and create a symlink to it. Then it gets out of the way. When you switch versions, it modifies the symlink to point to whichever Node.js installation you want. Then, again, it gets out of the way. 

Since npm is running and installing libraries in each of these installations, it proves NVM4W is working as intended.

The bottom line is I can confirm that these packages do install in both versions of Node.js, but Grunt, it's dependencies, and similar packages explicitly state they are not guaranteed to work in some of these versions.

Again, my advice is to determine which versions of grunt are designed for the versions of Node.js you want to run and install those explicitly, i.e. `npm install -g grunt@x.x.x` instead of `npm install -g grunt`. I don't think this is something NVM4W is causing and I don't think it is something that will provide further detail to troubleshoot this specific scenario.
